docs/sync-storage-handbook/src/places.md

```diff
@@ -0,0 +1,35 @@
+
+In the past, all Mozilla browsers handled history and bookmark storage differently. Desktop used its [own implementation, written in a mix of C++ and JavaScript](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Places); Firefox for Android had [one written in Java](https://searchfox.org/mozilla-central/rev/f1c7ba91fad60bfea184006f3728dd6ac48c8e56/mobile/android/base/java/org/mozilla/gecko/db/BrowserDB.java); and Firefox for iOS had a [third implementation in Swift](https://github.com/mozilla-mobile/firefox-ios/blob/ceb3cc9f0cc1ad8a90ad465b8c74a855121f319d/Storage/SQL/BrowserSchema.swift). All platforms used [SQLite](https://sqlite.org/) for storage, with a similarly-shaped schema, but the similarities ended there. Each implementation evolved independently, dictated by immediate product needs, and concepts like [frecency](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Places/Frecency_algorithm) and address autocomplete heuristics weren't shared at all. Each platform also had its own sync implementation. The goal of Rust Places is to unify these features across all platforms.
+
+Places is based on the Firefox Desktop implementation, and uses a mostly backward-compatible storage schema. It can either be consumed directly, as in Firefox for iOS, or through a layer like [Mozilla Android Components](https://mozac.org/).
+
+## Architecture
+
+`PlacesApi` is the entry point to Places. It's expected that your app or component will create one `PlacesApi` instance, and use it as a singleton. `PlacesApi` manages database connections and global Sync state. You can use `PlacesApi::open_connection` to open multiple read-only connections, or acquire the single read-write connection. Syncing for the first time also opens a special read-write connection for Sync, but this connection is not exposed to applications.
+
+`open_connection` returns a `PlacesDb`. This is a wrapper around a [Rusqlite](https://docs.rs/rusqlite) `Connection` with shared `PRAGMA`s, SQL functions, and support for interrupting long-running queries. Under the hood, `PlacesDb` supports encrypted (via [SQLCipher](https://www.zetetic.net/sqlcipher/)) and unencrypted SQLite databases.
+
+When you're done with a connection, make sure to close it using `PlacesApi::close_connection`.
+
+## Connection lifecycle
+
+Opening a read-write connection to an existing database runs migrations to bring the database up to date. The current schema and version are compiled in to the Rust Places library. Opening a database with a newer schema version is also supported, but the version will be rolled back to the current version. Opening a connection to a file that doesn't exist creates and initializes an empty database at that path.
```

Comment:
nit: opening a read-only connection also does the migrations using the read-write connection under the hood. Maybe this could say something like "Opening the first connection..." or similar?